File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/RetainGraphPresenter.java
package com.expensivebelly.dagger2retaingraph.activity;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class RetainGraphPresenter {

    private final Observable<String> timer;
    private RetainGraphView view;
    private Disposable disposable;

    public RetainGraphPresenter(Observable<String> timer) {
        this.timer = timer;
    }

    void attachView(RetainGraphView view) {
        this.view = view;
        startCounting();
    }

    private void startCounting() {
        disposable = timer.subscribe(new Consumer<String>() {
            @Override
            public void accept(String message) throws Exception {
                view.display(message);
            }
        });
    }

    void detachView() {
        view = new NullRetainGraphView();
        disposable.dispose();
    }
}

File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/RetainGraphPresenter.kt
package com.expensivebelly.dagger2retaingraph.activity

import com.expensivebelly.dagger2retaingraph.core.IPresenter
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.plusAssign

class RetainGraphPresenter(private val timer: Observable<String>) : IPresenter<RetainGraphView> {

    private val compositeDisposable = CompositeDisposable()

    override fun attach(view: RetainGraphView) {
        compositeDisposable += timer
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { message -> view.display(message) }
    }

    override fun detach() {
        compositeDisposable.clear()
    }
}